ejb3/src/main/java/org/jboss/as/ejb3/interceptor/server/ServerInterceptorCache.java

```diff
@@ -0,0 +1,161 @@
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.ejb3.interceptor.server;
+
+import org.jboss.as.ee.logging.EeLogger;
+import org.jboss.as.ee.utils.ClassLoadingUtils;
+import org.jboss.as.ejb3.component.ContainerInterceptorMethodInterceptorFactory;
+import org.jboss.as.ejb3.logging.EjbLogger;
+import org.jboss.as.naming.ManagedReference;
+import org.jboss.as.naming.ValueManagedReference;
+import org.jboss.invocation.InterceptorFactory;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.Index;
+import org.jboss.jandex.Indexer;
+import org.jboss.jandex.MethodInfo;
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleIdentifier;
+import org.jboss.modules.ModuleLoadException;
+import org.jboss.msc.value.CachedValue;
+import org.jboss.msc.value.ConstructedValue;
+import org.jboss.msc.value.Value;
+
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.AroundTimeout;
+import javax.interceptor.InvocationContext;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
+ */
+public class ServerInterceptorCache {
+
+    private static final Class<?>[] EMPTY_CLASS_ARRAY = new Class[0];
+
+    private final List<ServerInterceptorMetaData> serverInterceptorMetaData;
+
+    private List<InterceptorFactory> serverInterceptorsAroundInvoke = null;
+    private List<InterceptorFactory> serverInterceptorsAroundTimeout = null;
+
+    public ServerInterceptorCache(final List<ServerInterceptorMetaData> interceptorsMetaData){
+        this.serverInterceptorMetaData = interceptorsMetaData;
+    }
+
+    public List<InterceptorFactory> getServerInterceptorsAroundInvoke() {
+        synchronized(this) {
+            if (serverInterceptorsAroundInvoke == null) {
+                loadServerInterceptors();
```

Comment:
By calling this lazily instead of doing it in the constructor, you don't detect an invalid configuration during boot, but instead fail a subsequent deployment.

I guess that's ok though. One advantage of doing it this way is if in a managed domain this config ends up in the profile used by server groups that don't deploy EJB apps, and those servers don't have the necessary modules, the servers will be able to boot anyway.